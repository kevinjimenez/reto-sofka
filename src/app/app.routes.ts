import { Routes } from '@angular/router';
import { LayoutComponent } from './layout/layout.component';

export const routes: Routes = [
  {
    path: 'products',
    component: LayoutComponent,
    children: [
      {
        path: '',
        loadChildren: () =>
          import('./pages/products/products.routes').then((m) => m.routes),
      },
      {
        path: '',
        redirectTo: '',
        pathMatch: 'full',
      },
    ],
  },
  {
    path: '',
    redirectTo: 'products',
    pathMatch: 'full',
  },
  // {
  //   path: '**',
  //   loadComponent: () =>
  //     import('./pages/not-found/not-found.component').then(m => m.default),
  // },
];
