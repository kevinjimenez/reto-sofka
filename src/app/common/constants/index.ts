export { CONTEXT_MENU_OPTIONS } from './context-menu-options';
export { PRODUCT_TABLE_HEADERS } from './table-headers';
export { ERROR_INPUT_MESSAGE } from './error-input-message';
