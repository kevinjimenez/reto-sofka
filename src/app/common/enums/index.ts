export { MENU_OPTIONS } from './menu-options';
