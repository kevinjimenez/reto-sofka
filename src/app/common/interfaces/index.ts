export { Product } from './product.interface';
export { DefaultResponse } from './default-response.interface';
export { TableHeaderInterface } from './table-header.interface';
export { ErrorInputInterface } from './error-input.interface';
