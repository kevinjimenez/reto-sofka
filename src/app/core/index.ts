export { spinnerInterceptor } from './interceptors/spinner.interceptor';
export { productsResolver } from './resolvers/products.resolver';
export { ProductsService } from './services/products.service';
export { SpinnerService } from './services/spinner.service';
