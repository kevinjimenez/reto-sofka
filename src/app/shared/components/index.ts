export { ContextMenuOption } from './context-menu/interfaces/context-menu-option.interface';
export { ContextMenuOptionSelected } from './context-menu/interfaces/context-menu-option-selected.interface';
export { ContextMenuComponent } from './context-menu/context-menu.component';
export { IconButtonComponent } from './icon-button/icon-button.component';
export { AlertCircleComponent } from './icons/alert-circle/alert-circle.component';
export { OverlayComponent } from './overlay/overlay.component';
export { SpinnerComponent } from './spinner/spinner.component';
export { ButtonComponent } from './button/button.component';
export { InputComponent } from './input/input.component';
export { LogoComponent } from './logo/logo.component';
export { ToastComponent } from './toast/toast.component';
export { TooltipComponent } from './tooltip/tooltip.component';
