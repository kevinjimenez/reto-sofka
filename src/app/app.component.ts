import { Component } from '@angular/core';
import { RouterOutlet } from '@angular/router';
import { SpinnerComponent } from './shared/components';

@Component({
	selector: 'app-root',
	standalone: true,
	imports: [RouterOutlet, SpinnerComponent],
	templateUrl: './app.component.html',
	styleUrl: './app.component.css'
})
export class AppComponent {}
