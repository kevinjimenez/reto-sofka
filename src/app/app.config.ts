import { ApplicationConfig, provideZoneChangeDetection } from '@angular/core';
import { provideRouter, withComponentInputBinding, withHashLocation } from '@angular/router';

import { provideHttpClient, withInterceptors } from '@angular/common/http';
import { routes } from './app.routes';
import { spinnerInterceptor } from './core';

export const appConfig: ApplicationConfig = {
	providers: [
		provideZoneChangeDetection({ eventCoalescing: true }),
		provideHttpClient(withInterceptors([spinnerInterceptor])),
		provideRouter(
			routes,
			withHashLocation(),
			withComponentInputBinding()
			// withPreloading(PreloadAllModules)
		)
	]
};
