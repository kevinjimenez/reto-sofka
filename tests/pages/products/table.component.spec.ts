import { ComponentRef } from '@angular/core';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Router } from '@angular/router';
import { of, throwError } from 'rxjs';
import { MENU_OPTIONS } from '../../../src/app/common/enums';
import { Product } from '../../../src/app/common/interfaces';
import { ProductsService } from '../../../src/app/core';
import { TableComponent } from '../../../src/app/pages/products/components/table/table.component';
import { ContextMenuOptionSelected } from '../../../src/app/shared/components';

const mockProduct: Product = {
	id: '1',
	name: 'Product 1',
	description: 'Description 1',
	logo: 'logo1.png',
	date_release: '2024-01-01',
	date_revision: '2024-01-01'
};

describe('TableComponent', () => {
	let component: TableComponent;
	let componentRef: ComponentRef<TableComponent>;
	let fixture: ComponentFixture<TableComponent>;
	let compiled: HTMLElement;
	let productsServiceMock = { deleteById: jest.fn() };
	let routerMock = { navigate: jest.fn() };

	beforeEach(async () => {
		await TestBed.configureTestingModule({
			imports: [TableComponent],
			providers: [
				{ provide: ProductsService, useValue: productsServiceMock },
				{ provide: Router, useValue: routerMock }
			]
		}).compileComponents();
	});

	beforeEach(() => {
		fixture = TestBed.createComponent(TableComponent);
		component = fixture.componentInstance;
		componentRef = fixture.componentRef;
		fixture.detectChanges();
		compiled = fixture.nativeElement;
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should match snapshot', () => {
		expect(compiled).toMatchSnapshot();
	});

	describe('optionSelected', () => {
		it('should navigate to product edit page when update is selected', () => {
			component.itemSelected.set(mockProduct);
			component.optionSelected({ option: MENU_OPTIONS.UPDATE } as ContextMenuOptionSelected);
			expect(routerMock.navigate).toHaveBeenCalledWith(['products', 'product', mockProduct.id], {
				state: { payload: mockProduct }
			});
		});

		it('should navigate to product delete page when delete is selected', () => {
			component.itemSelected.set(mockProduct);
			component.optionSelected({ option: MENU_OPTIONS.DELETE } as ContextMenuOptionSelected);

			expect(component.overlayVisible()).toBe(true);
		});
	});

	describe('onDelete', () => {
		it('should call deleteById and emit onRemoveItem when onDelete is called', () => {
			component.itemSelected.set(mockProduct);
			productsServiceMock.deleteById.mockReturnValue(of({ message: 'Product deleted' } as any));
			const onRemoveItemSpy = jest.spyOn(component.onRemoveItem, 'emit');
			component.onDelete();

			expect(productsServiceMock.deleteById).toHaveBeenCalledWith(mockProduct.id);
			expect(onRemoveItemSpy).toHaveBeenCalledWith(mockProduct.id);
		});

		it('should trhow error when deleteById', () => {
			const errorMessage = 'Error deleting product';
			productsServiceMock.deleteById.mockReturnValue(
				throwError(() => {
					return { error: { message: errorMessage } };
				})
			);

			component.onDelete();

			expect(productsServiceMock.deleteById).toHaveBeenCalledWith('');

			expect(component.errorMsg()).toBe(errorMessage);
			expect(component.toastVisible()).toBe(true);
		});
	});

	describe('onSelectionChange', () => {
		it('should emit onViewItems when onSelectionChange is called', () => {
			const onViewItemsSpy = jest.spyOn(component.onViewItems, 'emit');
			const newValue = '10';
			component.onSelectionChange(newValue);

			expect(onViewItemsSpy).toHaveBeenCalledWith(10);
		});
	});

	describe('showContextMenu', () => {
		it('should show context menu', () => {
			componentRef.setInput('products', [mockProduct]);
			fixture.detectChanges();

			const event = new MouseEvent('click');
			component.showContextMenu(event, mockProduct);

			expect(component.contextMenuVisible()).toBe(true);
		});
	});
});
