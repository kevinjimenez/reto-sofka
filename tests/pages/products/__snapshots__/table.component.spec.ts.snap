// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <app-toast>
    
  </app-toast>
  <section
    class="table-container"
  >
    <div
      class="table-wrapper"
    >
      <table>
        <thead>
          <tr>
            <th>
              <section
                class="flex items-center w-full gap-1/2 cursor-default"
              >
                 Logo 
                
              </section>
            </th>
            <th>
              <section
                class="flex items-center w-full gap-1/2 cursor-default"
              >
                 Nombre del producto 
                
              </section>
            </th>
            <th>
              <section
                class="flex items-center w-full gap-1/2 cursor-default"
              >
                 Descripción 
                <span
                  class="tooltip-container"
                >
                  <app-alert-circle
                    class="cursor-pointer"
                  >
                    <svg
                      class="primary"
                      fill="currentColor"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                        stroke="none"
                      />
                      <path
                        d="M12 2c5.523 0 10 4.477 10 10a10 10 0 0 1 -19.995 .324l-.005 -.324l.004 -.28c.148 -5.393 4.566 -9.72 9.996 -9.72zm.01 13l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-.01 -8a1 1 0 0 0 -.993 .883l-.007 .117v4l.007 .117a1 1 0 0 0 1.986 0l.007 -.117v-4l-.007 -.117a1 1 0 0 0 -.993 -.883z"
                      />
                    </svg>
                  </app-alert-circle>
                  <app-tooltip>
                    <main
                      class="tooltip-content"
                    >
                       Descripción

                    </main>
                  </app-tooltip>
                </span>
                
              </section>
            </th>
            <th>
              <section
                class="flex items-center w-full gap-1/2 cursor-default"
              >
                 Fecha de liberación 
                <span
                  class="tooltip-container"
                >
                  <app-alert-circle
                    class="cursor-pointer"
                  >
                    <svg
                      class="primary"
                      fill="currentColor"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                        stroke="none"
                      />
                      <path
                        d="M12 2c5.523 0 10 4.477 10 10a10 10 0 0 1 -19.995 .324l-.005 -.324l.004 -.28c.148 -5.393 4.566 -9.72 9.996 -9.72zm.01 13l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-.01 -8a1 1 0 0 0 -.993 .883l-.007 .117v4l.007 .117a1 1 0 0 0 1.986 0l.007 -.117v-4l-.007 -.117a1 1 0 0 0 -.993 -.883z"
                      />
                    </svg>
                  </app-alert-circle>
                  <app-tooltip>
                    <main
                      class="tooltip-content"
                    >
                       Fecha de liberación

                    </main>
                  </app-tooltip>
                </span>
                
              </section>
            </th>
            <th>
              <section
                class="flex items-center w-full gap-1/2 cursor-default"
              >
                 Fecha de reestructuración 
                <span
                  class="tooltip-container"
                >
                  <app-alert-circle
                    class="cursor-pointer"
                  >
                    <svg
                      class="primary"
                      fill="currentColor"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        fill="none"
                        stroke="none"
                      />
                      <path
                        d="M12 2c5.523 0 10 4.477 10 10a10 10 0 0 1 -19.995 .324l-.005 -.324l.004 -.28c.148 -5.393 4.566 -9.72 9.996 -9.72zm.01 13l-.127 .007a1 1 0 0 0 0 1.986l.117 .007l.127 -.007a1 1 0 0 0 0 -1.986l-.117 -.007zm-.01 -8a1 1 0 0 0 -.993 .883l-.007 .117v4l.007 .117a1 1 0 0 0 1.986 0l.007 -.117v-4l-.007 -.117a1 1 0 0 0 -.993 -.883z"
                      />
                    </svg>
                  </app-alert-circle>
                  <app-tooltip>
                    <main
                      class="tooltip-content"
                    >
                       Fecha de reestructuración

                    </main>
                  </app-tooltip>
                </span>
                
              </section>
            </th>
            <th>
              <section
                class="flex items-center w-full gap-1/2 cursor-default"
              >
                  
                
              </section>
            </th>
            
          </tr>
        </thead>
        <tbody>
          
          <td
            colspan="5"
          >
            <section
              class="not-records"
            >
              Sin registros
            </section>
          </td>
          
        </tbody>
      </table>
    </div>
    <footer
      class="table-footer"
    >
      <span>
        0 de 0 Resultados
      </span>
      <select
        class="ng-untouched ng-pristine ng-valid"
      >
        <option
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="15"
        >
          15
        </option>
        
      </select>
    </footer>
  </section>
  <app-overlay>
    
  </app-overlay>
</div>
`;
